File: leetcode/Remove Duplicates from Sorted List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (NULL == head)
        {
			return head;
		}
        ListNode *parent = head, *current = head->next;
        int current_val = head->val;
        while (current)
        {
			if (current->val == current_val)
			{
				parent->next = current->next;
				delete current;
				current = parent;
			} else
			{
				current_val = current->val;
			}
			// if the current is deleted, the parent does not move
			if (NULL != parent && current != parent)
			{
				parent = parent->next;
			}
			if (NULL != current)
			{
				current = current->next;
			}
		}
		return head;
    }
};

File: answers/leetcode/Remove Duplicates from Sorted List/Remove Duplicates from Sorted List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """

        if None == head or None == head.next:
            return head
        elif head.val == head.next.val:
            return self.deleteDuplicates(head.next)
        else:
            head.next = self.deleteDuplicates(head.next)
            return head
